src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs

```diff
@@ -0,0 +1,658 @@
+                        _requestWakeupPipeFd = fds[Interop.Sys.WriteEndOfPipe];
+                    }
+
+                    // Mark the worker as running.
+                    _workerRunning = true;
+
+                    // Kick off the processing task.  It's "DenyChildAttach" to avoid any surprises if
+                    // code happens to create attached tasks, and it's LongRunning because this thread
+                    // is lkely going to sit around for a while in a wait loop (and the more requests
+                    // are concurrently issued to the same agent, the longer the thread will be around).
+                    const TaskCreationOptions Options = TaskCreationOptions.DenyChildAttach | TaskCreationOptions.LongRunning;
+                    Task.Factory.StartNew(s =>
+                    {
+                        var thisRef = (MultiAgent)s;
+                        try
+                        {
+                            // Do the actual processing
+                            thisRef.Process();
+                        }
+                        catch (Exception exc)
+                        {
+                            Debug.Fail("Unexpected exception from processing loop: " + exc.ToString());
+                        }
+                        finally
+                        {
+                            lock (thisRef._incomingRequests)
+                            {
+                                // Close our wakeup pipe (ignore close errors).
+                                // This is done while holding the lock to prevent
+                                // subsequent Queue calls to see an improperly configured
+                                // set of descriptors.
+                                Interop.Sys.Close(thisRef._wakeupRequestedPipeFd);
+                                Interop.Sys.Close(thisRef._requestWakeupPipeFd);
+                                thisRef._wakeupRequestedPipeFd = 0;
+                                thisRef._requestWakeupPipeFd = 0;
+
+                                // In the time between we stopped processing and now,
+                                // more requests could have been added.  If they were
+                                // kick off another processing loop.
+                                thisRef._workerRunning = false;
+                                if (thisRef._incomingRequests.Count > 0)
+                                {
+                                    thisRef.EnsureWorkerIsRunning();
+                                }
+                            }
+                        }
+                    }, this, CancellationToken.None, Options, TaskScheduler.Default);
+                }
+                else // _workerRunning == true
+                {
+                    // The worker is already running.  If there are already queued requests, we're done.
+                    // However, if there aren't any queued requests, Process could be blocked inside of
```

Comment:
> Wouldn't it return under the FailSafeTimeOut?

Yes, but that timeout is long (and we potentially might want to make it longer).  That timeout is really there as a "failsafe", just in case something goes wrong, but really shouldn't be necessary.  This mechanism is meant to be the primary way that we wake up the waiting.  If we didn't have this, we'd need to make the timeout much shorter, in which case we'd have a thread polling and doing a busy waiting loop, e.g. waking up 30 times a second to see if there's anything to be done.  I currently think this strikes a nice compromise, but we can always revisit it later.